airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py

```diff
@@ -0,0 +1,219 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+
+from copy import deepcopy
+from dataclasses import InitVar, dataclass
+from typing import Any, List, Mapping, MutableMapping, Optional, Union
+
+import dpath
+from deprecated.classic import deprecated
+
+from airbyte_cdk.sources.declarative.interpolation.interpolated_string import InterpolatedString
+from airbyte_cdk.sources.declarative.retrievers.retriever import Retriever
+from airbyte_cdk.sources.declarative.schema.schema_loader import SchemaLoader
+from airbyte_cdk.sources.source import ExperimentalClassWarning
+from airbyte_cdk.sources.types import Config
+
+AIRBYTE_DATA_TYPES: Mapping[str, Mapping[str, Any]] = {
+    "string": {"type": ["null", "string"]},
+    "boolean": {"type": ["null", "boolean"]},
+    "date": {"type": ["null", "string"], "format": "date"},
+    "timestamp_without_timezone": {
+        "type": ["null", "string"],
+        "format": "date-time",
+        "airbyte_type": "timestamp_without_timezone",
+    },
+    "timestamp_with_timezone": {"type": ["null", "string"], "format": "date-time"},
+    "time_without_timezone": {
+        "type": ["null", "string"],
+        "format": "time",
+        "airbyte_type": "time_without_timezone",
+    },
+    "time_with_timezone": {
+        "type": ["null", "string"],
+        "format": "time",
+        "airbyte_type": "time_with_timezone",
+    },
+    "integer": {"type": ["null", "integer"]},
+    "number": {"type": ["null", "number"]},
+    "array": {"type": ["null", "array"]},
+    "object": {"type": ["null", "object"]},
+}
+
+
+@deprecated("This class is experimental. Use at your own risk.", category=ExperimentalClassWarning)
+@dataclass(frozen=True)
+class TypesMap:
+    """
+    Represents a mapping between a current type and its corresponding target type.
+    """
+
+    target_type: Union[List[str], str]
+    current_type: Union[List[str], str]
+
+
+@deprecated("This class is experimental. Use at your own risk.", category=ExperimentalClassWarning)
+@dataclass
+class SchemaTypeIdentifier:
+    """
+    Identifies schema details for dynamic schema extraction and processing.
+    """
+
+    key_pointer: List[Union[InterpolatedString, str]]
+    parameters: InitVar[Mapping[str, Any]]
+    type_pointer: Optional[List[Union[InterpolatedString, str]]] = None
+    types_map: Optional[List[TypesMap]] = None
```

Comment:
nit: Should this be named `types_mapping` like in the yaml and model?

```
      types_mapping:
        type: array
        items:
          - "$ref": "#/definitions/TypesMap"
```

Reply:
updated